fastparquet/test/test_writer.py

```diff
@@ -506,3 +506,141 @@ def test_many_categories(tempdir, n):
+    assert len(pf.row_groups) == 2 * len(row_groups)
+    out = pf.to_pandas().a.tolist()
+    assert out == [1, 2, 3, 4, 5, 6, 7, 8]
+    if "cat=1" in dirs:
+        assert 'cat' in pf.cats
+
+
+def test_merge_fail(tempdir):
+    fn = str(tempdir)
+
+    df0 = pd.DataFrame({'a': [1, 2, 3, 4]})
+    fn0 = os.sep.join([fn, 'out0.parq'])
+    write(fn0, df0)
+
+    df1 = pd.DataFrame({'a': ['a', 'b', 'c']})
+    fn1 = os.sep.join([fn, 'out1.parq'])
+    write(fn1, df1)
+
+    with pytest.raises(ValueError) as e:
+        writer.merge([fn0, fn1])
+    assert 'schemas' in str(e)
+
+    os.remove(fn1)
+    write(fn1, df0, file_scheme='hive')
+    with pytest.raises(ValueError) as e:
+        writer.merge([fn0, fn1])
+    assert 'multi-file' in str(e)
+
+
+def test_analyse_paths():
+    file_list = ['a', 'b']
+    base, out = writer.analyse_paths(file_list, '/')
+    assert (base, out) == ('', ['a', 'b'])
+
+    file_list = ['c/a', 'c/b']
+    base, out = writer.analyse_paths(file_list, '/')
+    assert (base, out) == ('c', ['a', 'b'])
+
+    file_list = ['c/d/a', 'c/d/b']
+    base, out = writer.analyse_paths(file_list, '/')
+    assert (base, out) == ('c/d', ['a', 'b'])
+
+    file_list = ['c/cat=1/a', 'c/cat=2/b', 'c/cat=1/c']
+    base, out = writer.analyse_paths(file_list, '/')
+    assert (base, out) == ('c', ['cat=1/a', 'cat=2/b', 'cat=1/c'])
+
+    file_list = ['c/cat=2/b', 'c/cat/a', 'c/cat=1/c']
+    with pytest.raises(ValueError) as e:
+        writer.analyse_paths(file_list, '/')
+    assert 'c/cat/a' in str(e)
+
+    file_list = ['c/cat=2/b', 'c/fred=2/a', 'c/cat=1/c']
+    with pytest.raises(ValueError) as e:
+        writer.analyse_paths(file_list, '/')
+    assert 'directories' in str(e)
+
+    file_list = ['c/cat=2/b', 'c/a', 'c/cat=1/c']
+    with pytest.raises(ValueError) as e:
+        writer.analyse_paths(file_list, '/')
+    assert 'nesting' in str(e)
+
+
+def test_append_simple(tempdir):
+    fn = os.path.join(str(tempdir), 'test.parq')
+    df = pd.DataFrame({'a': [1, 2, 3, 0],
+                       'b': ['a', 'a', 'b', 'b']})
+    write(fn, df, write_index=False)
+    write(fn, df, append=True, write_index=False)
+
+    pf = ParquetFile(fn)
+    expected = pd.concat([df, df], ignore_index=True)
+    pd.util.testing.assert_frame_equal(pf.to_pandas(), expected)
+
+
+@pytest.mark.parametrize('row_groups', ([0], [0, 2]))
+@pytest.mark.parametrize('partition', ([], ['b']))
+def test_append(tempdir, row_groups, partition):
+    fn = str(tempdir)
+    df0 = pd.DataFrame({'a': [1, 2, 3, 0],
+                        'b': ['a', 'b', 'a', 'b'],
+                        'c': True})
+    df1 = pd.DataFrame({'a': [4, 5, 6, 7],
+                        'b': ['a', 'b', 'a', 'b'],
+                        'c': False})
+    write(fn, df0, partition_on=partition, file_scheme='hive',
+          row_group_offsets=row_groups)
+    write(fn, df1, partition_on=partition, file_scheme='hive',
+          row_group_offsets=row_groups, append=True)
```

Comment:
`test_append_simple`, above, does this (since "simple" file scheme is the default). It should succeed, except if the file-system doesn't allow 'rb+' mode. That test is not made, there is no test in append/merge yet for s3.